hngprojects/hng_boilerplate_nestjs
# feat: updating a blog by superadmin

---

## What does the PR do?

This PR introduces a new endpoint for updating blog posts that is restricted to super admins. The endpoint only allows access if the provided token is valid and has the super admin role. The PR ensures that:

- The endpoint handles missing or invalid authorization headers with a 401 status code.
- It returns a 403 status code if the token does not have the super admin role.
- It correctly validates input fields and responds with a 400 status code for invalid requests.
- If the blog post with the provided ID is not found, it returns a 404 status code.
- On successful blog update, it returns a 200 status code with the appropriate response body.

## How should this be manually tested?

1. Ensure you have a valid super admin token.
2. Send a PUT request to the following endpoint:

   ```
   PUT /api/v1/blogs/:id
   ```

   Replace `:id` with the ID of the blog post you want to update.
   
3. Include the Authorization header:

   ```
   Authorization: Bearer <access_token>
   ```

4. Provide a valid request body with the updated blog details:

   ```json
   {
     "title": "Updated Blog Title",
     "content": "<p>Updated first paragraph.</p><p>Updated second paragraph.</p>",
     "tags": ["updated-tag1", "updated-tag2"],
     "image_urls": ["https://example.com/updated-image1.jpg", "https://example.com/updated-image2.jpg"]
   }
   ```

## Success Response

On a successful execution, you will receive the following response structure:

### blog_update_success

```json
{
  "status_code": 200,
  "message": "Blog post updated successfully",
  "post": {
    "blog_id": "unique-blog-id",
    "title": "Updated Blog Title",
    "content": "<p>Updated first paragraph.</p><p>Updated second paragraph.</p>",
    "image_urls": ["https://example.com/updated-image1.jpg", "https://example.com/updated-image2.jpg"],
    "tags": ["updated-tag1", "updated-tag2"],
    "author": "Super Admin",
    "created_at": "2024-08-05T19:03:52.194Z"
  }
}
```

**Example Response Image:**
![updateblogs-succes](https://github.com/user-attachments/assets/9ec23010-044e-4571-b6d6-51f37a9ddd47)

## Unauthorized Response

If the token is missing or invalid, you will receive the following response:

### blog_unauthorized

```json
{
  "status_code": 401,
  "message": "Unauthorized. Please log in."
}
```


## "Not Found" Response

If the blog post with the provided ID is not found, you will receive the following response:

### blog_not_found

```json
{
  "status_code": 404,
  "message": "Blog post not found."
}
```


## "Internal Server Error" Response

For unexpected server issues, you will receive the following response:

### blog_internal_server_error

```json
{
  "status_code": 500,
  "message": "Internal server error. Failed to update the blog post."
}
```

## Task

- Implemented PUT /api/v1/blogs/:id endpoint in `blog.controller.ts`.
- Added method in `BlogService` to handle blog updates with super admin restriction.
- Validated token and user role to ensure only super admins can update blogs.
- Implemented appropriate error handling for invalid tokens, request bodies, and non-existent blog posts.
- Added unit tests for the endpoint covering authorization, input validation, and success cases.
- Updated API documentation to reflect the new endpoint and its requirements.

## Unit Test Evidence

**blog_update_test**
![update-blog-unit-test](https://github.com/user-attachments/assets/97fcd3aa-471a-4464-a6bd-886d668c44e1)


Reference Issue: [Link to Issue](https://linear.app/team-bingo/issue/BAC2-127/endpoint-to-edit-blog-post-or-136)

---
